File: tests/mem.zig
const std = @import("std");
const vortex = @import("vortex");

const mem = vortex.mem;
const testing = std.testing;

test "AllocatorDuplicate" {
    var pages1 = try mem.PageAllocator.init(2);
    defer pages1.deinit();
    var arena1 = mem.Allocator.init(pages1.mem, .arena);
    _ = try arena1.alloc(u128, 12);
    _ = try arena1.alloc(u64, 14);

    var pages2 = try mem.PageAllocator.init(4);
    defer pages2.deinit();
    var arena2 = mem.Allocator.init(pages2.mem, .arena);

    const duplicate = try arena2.dupe(u8, pages1.mem);
    try testing.expect(mem.eql(u8, pages1.mem, duplicate));
}

test "PageAllocator" {
    var pages = try mem.PageAllocator.init(84);
    defer pages.deinit();

    try testing.expect(pages.mem.len == vortex.os.page_size * 84);
}

test "ArenaAllocator" {
    var pages = try mem.PageAllocator.init(84);
    defer pages.deinit();
    var arena = mem.Allocator.init(pages.mem, .arena);

    _ = try arena.alloc(u128, 64);
    _ = try arena.alloc(u64, 54);

    try testing.expect(arena.ctx.arena.pos ==
        (vortex.math.alignForward(usize, @sizeOf(u128) * 64, 8) +
        vortex.math.alignForward(usize, @sizeOf(u64) * 54, 8)));
}

test "FrelistAllocator" {
    var pages = try mem.PageAllocator.init(1);
    defer pages.deinit();
    var freelist = mem.Allocator.init(pages.mem, .freelist);

    try testing.expect(freelist.ctx.freelist.freelist == null);

    const allocated1 = try freelist.alloc(u8, 64);
    const allocated2 = try freelist.alloc(u128, 64);
    freelist.free(allocated1);
    freelist.free(allocated2);

    try testing.expect(freelist.ctx.freelist.freelist != null);
    try testing.expect(freelist.ctx.freelist.freelist.?.next != null);
    try testing.expect(freelist.ctx.freelist.freelist.?.len == @sizeOf(u128) * 64);
    try testing.expect(freelist.ctx.freelist.freelist.?.next.?.len == @sizeOf(u8) * 64);
}

test "ArrayList" {
    var pages = try mem.PageAllocator.init(5);
    defer pages.deinit();
    var arena = mem.Allocator.init(pages.mem, .arena);
    var arraylist = try mem.ArrayList(usize).init(&arena);

    inline for (0..30) |idx| {
        try testing.expect(arraylist.pos == idx);
        try arraylist.push(idx);
    }

    inline for (0..30) |idx| {
        try testing.expect(arraylist.get(idx).? == idx);
    }
}

File: tests/mem.cpp
#define NO_ARGS true

#include <vortex/vortex.hpp>

template <typename T> static bool is_aligned(const T *ptr) {
    return (reinterpret_cast<usize>(ptr) % alignof(T)) == 0;
}

static vortex::SysRes<vortex::None> main_w() {
    vortex::PageAllocator p = TRY(vortex::PageAllocator::init(2));
    vortex::assert(p.len == 2 * PAGE_SIZE, "Invalid number of allocated pages!");

    vortex::Allocator<vortex::Arena> a = vortex::Allocator<vortex::Arena>::init(p.len, p.ptr);

    const vortex::Slice<u8> m1         = a.alloc<u8>(25);
    vortex::assert(m1.ptr != NULL, "Failed to allocate memory from the Arena!");
    vortex::assert(is_aligned<u8>(m1.ptr), "Unaligned memory!");
    vortex::assert(a.state.pos == 25 * sizeof(u8), "Invalid Arena position!");

    const vortex::Slice<vortex::PageAllocator> m2 = a.alloc<vortex::PageAllocator>(172);
    vortex::assert(m2.ptr != NULL, "Failed to allocate memory from the Arena!");
    vortex::assert(is_aligned<vortex::PageAllocator>(m2.ptr), "Unaligned memory!");
    vortex::assert(
        a.state.pos == vortex::alignForward(
                           (25 * sizeof(u8)) + (172 * sizeof(vortex::PageAllocator)),
                           alignof(vortex::PageAllocator)
                       ),
        "Invalid Arena position!"
    );

    const vortex::Slice<float> m3 = a.alloc<float>(172);
    vortex::assert(m3.ptr != NULL, "Failed to allocate memory from the Arena!");
    vortex::assert(is_aligned<float>(m3.ptr), "Unaligned memory!");

    const vortex::Slice<long double> m4 = a.alloc<long double>(172);
    vortex::assert(m4.ptr != NULL, "Failed to allocate memory from the Arena!");
    vortex::assert(is_aligned<long double>(m4.ptr), "Unaligned memory!");

    p.deinit();

    return vortex::None();
}

static u8 main() {
    vortex::SysRes<vortex::None> res = main_w();
    if (res.is_err()) {
        vortex::println("Error: ", static_cast<u8>(res.kind));
        vortex::assert(false, "Failed to run the main function!");

        return 1;
    }

    return 0;
}
